tests/subsys/net/lib/azure_iot_hub/src/main.c

```diff
@@ -0,0 +1,234 @@
+	struct topic_parser_data topic_data = {
+		.topic = topic,
+		.topic_len = strlen(topic),
+		.type = TOPIC_TYPE_UNKNOWN,
+	};
+
+	err = azure_iot_hub_topic_parse(&topic_data);
+	zassert_equal(err, 0, NULL);
+	zassert_equal(topic_data.type, TOPIC_TYPE_TWIN_UPDATE_DESIRED,
+		      "Incorrect topic type");
+	zassert_equal(topic_data.prop_bag_count, 0,
+		      "Incorrect property bag count");
+}
+
+static void test_topic_parse_twin_update_result(void)
+{
+	int err;
+	const char *topic = "$iothub/twin/res/200/?$rid=738&$version=135";
+	struct topic_parser_data topic_data = {
+		.topic = topic,
+		.topic_len = strlen(topic),
+		.type = TOPIC_TYPE_UNKNOWN,
+	};
+
+	err = azure_iot_hub_topic_parse(&topic_data);
+	zassert_equal(err, 0, NULL);
+	zassert_equal(topic_data.type, TOPIC_TYPE_TWIN_UPDATE_RESULT,
+		      "Incorrect topic type");
+	zassert_equal(topic_data.status, 200, NULL);
+	zassert_equal(topic_data.prop_bag_count, 2,
+		      "Incorrect property bag count");
+	zassert_true(strcmp(topic_data.prop_bag[0].key, "$rid") == 0, NULL);
+	zassert_true(strcmp(topic_data.prop_bag[0].value, "738") == 0, NULL);
+	zassert_true(strcmp(topic_data.prop_bag[1].key, "$version") == 0, NULL);
+	zassert_true(strcmp(topic_data.prop_bag[1].value, "135") == 0, NULL);
+}
+
+static void test_topic_parse_direct_method(void)
+{
+	int err;
+	const char *topic = "$iothub/methods/POST/my_method/?$rid=387";
+	struct topic_parser_data topic_data = {
+		.topic = topic,
+		.topic_len = strlen(topic),
+		.type = TOPIC_TYPE_UNKNOWN,
+	};
+
+	err = azure_iot_hub_topic_parse(&topic_data);
+	zassert_equal(err, 0, NULL);
+	zassert_equal(topic_data.type, TOPIC_TYPE_DIRECT_METHOD,
+		      "Incorrect topic type");
+	zassert_true(strcmp(topic_data.name, "my_method") == 0, NULL);
+	zassert_equal(topic_data.prop_bag_count, 1,
+		      "Incorrect property bag count");
+	zassert_true(strcmp(topic_data.prop_bag[0].key, "$rid") == 0, NULL);
+	zassert_true(strcmp(topic_data.prop_bag[0].value, "387") == 0, NULL);
+}
+
+static void test_topic_parse_dps_reg_result(void)
+{
+	int err;
+	const char *topic =
+		"$dps/registrations/res/204/?$rid=59765&retry-after=3";
+	struct topic_parser_data topic_data = {
+		.topic = topic,
+		.topic_len = strlen(topic),
+		.type = TOPIC_TYPE_UNKNOWN,
+	};
+
+	err = azure_iot_hub_topic_parse(&topic_data);
+	zassert_equal(err, 0, NULL);
+	zassert_equal(topic_data.type, TOPIC_TYPE_DPS_REG_RESULT,
+		      "Incorrect topic type");
+	zassert_equal(topic_data.status, 204, NULL);
+	zassert_equal(topic_data.prop_bag_count, 2,
+		      "Incorrect property bag count");
+	zassert_true(strcmp(topic_data.prop_bag[0].key, "$rid") == 0, NULL);
+	zassert_true(strcmp(topic_data.prop_bag[0].value, "59765") == 0, NULL);
+	zassert_true(strcmp(topic_data.prop_bag[1].key, "retry-after") == 0,
+			    NULL);
+	zassert_true(strcmp(topic_data.prop_bag[1].value, "3") == 0, NULL);
+}
+
+static void test_topic_parse_prop_bag_overload(void)
```

Comment:
The test key/value data for property bags are pure ascii, however they are RFC 2396-encoded values, so this should also work as key and/or value: `det%20skal%20ogs%C3%A5%20funker%20%F0%9F%98%87` and parsed into `det skal også funker 😇` (however I don't know if Zephyr supports unicode strings).

Same goes for adding values.

Please extend the tests with some more complicated examples.

Reply:
Yeah, that's why the PR is a draft. I haven't figured out a good way to solve this. It adds quite a bit of complexity to the implementation.